### PR TITLE
cjdns: Improving systemd unit description

### PR DESCRIPTION
**Pull request by spacekitteh**
###### Motivation for this change
Makes cjdns actually require networking.

### CODE REVIEW

**Comment by groxxda** [comment]
Why is the bindsTo needed?

**Comment by spacekitteh** [comment]
Because it requires networking to be up before it can create a connection. If you have a slow-starting network connection, then cjdns will fail if it's started too soon after starting networking.

If networking goes down, then having cjdns running is pointless; I'm also not sure if it can even re-establish a connection after losing the network yet.

**Comment by groxxda** [comment]
`networking.target` gives no guarantee that networking is _up_. You might want to depend on `network-online.target`

**Comment by groxxda** [comment]
```
systemctl list-dependencies sleep.target --reverse --type=target --all 
sleep.target
● ├─systemd-hibernate.service
● │ └─hibernate.target
● ├─systemd-hybrid-sleep.service
● │ └─hybrid-sleep.target
● └─systemd-suspend.service
●   └─suspend.target
```
Depending on `sleep.target` should suffice, since the other three will pull in `sleep.target`

**Comment by spacekitteh** [comment]
I guess I could just disable start rate limiting. I'll give that a try.

**Comment by spacekitteh** [comment]
Well, it works now. \o/ I squashed the commits.

**Review by MostAwesomeDude — APPROVED** [review]
LGTM

I haven't tested but the submitter says that it works for them, so that works for me.

**Comment by grahamc** [comment]
@groxxda what do you think of the current state?

**Comment by groxxda** [comment]
I don't use cjdns so I cannot say anything about the issues @spacekitteh experienced.
At least the upstream service file does not require network-online, and I don't understand why this is needed here.

I don't know why ordering after network-online does not work, that may be a bug in the networking module. Of course I'd favor fixing the real issue and not working around it in clients with unlimited restarts, but that's just my opinion.
If it was broken before and works like this, it's clearly an improvement..

**Comment by spacekitteh** [comment]
Is it good to merge?

**Comment by edolstra** [comment]
Note my comment about systemd unit descriptions.

**Comment by spacekitteh** [comment]
@edolstra Where?

File: nixos/modules/services/networking/cjdns.nix [hunk 1]
```diff
@@ -204,11 +204,28 @@ in
...
+      wantedBy = ["sleep.target"];
+      bindsTo = ["network-online.target"];
+      after = ["sleep.target" "network-online.target"];
+
+      script = 
+        "
+          ${config.systemd.package}/bin/systemctl systemctl reset-failed cjdns.service
+          ${config.systemd.package}/bin/systemctl restart cjdns.service
+        ";
+      postStop = "${config.systemd.package}/bin/systemctl systemctl reset-failed cjdns-restart.service";
```
Comment (Mic92) [inline nixos/modules/services/networking/cjdns.nix]: `systemctl systemctl` is wrong in:

```
${config.systemd.package}/bin/systemctl systemctl reset-failed cjdns-restart.service
```

File: nixos/modules/services/networking/cjdns.nix [hunk 2]
```diff
@@ -204,11 +204,28 @@ in
...
-
+    systemd.services.cjdns-restart = {
+      description = "cjdns: restart after resuming from sleep, hibernation, hybrid-sleep and hybrid-suspend; and other failures";
+      wantedBy = ["sleep.target"];
+      bindsTo = ["network-online.target"];
+      after = ["sleep.target" "network-online.target"];
+
+      script = 
+        "
+          ${config.systemd.package}/bin/systemctl systemctl reset-failed cjdns.service
```
Comment (Mic92) [inline nixos/modules/services/networking/cjdns.nix]: same as below

File: nixos/modules/services/networking/cjdns.nix [hunk 3]
```diff
@@ -241,10 +258,13 @@ in
                 | ${pkg}/bin/cjdroute
          ''
       );
-
+      
+      onFailure = ["cjdns-restart.service"];
```
Comment (Mic92) [inline nixos/modules/services/networking/cjdns.nix]: Why is that needed?

**Comment by Mic92** [comment]
I don't think an attacker can perform an MITM: cjdns does end-to-end encryption based on addresses.

**Comment by Mic92** [comment]
@cjdelisle is it possible to implement a signal handler for SIGUSR2 to just reconnect to its peers? That would make the cjdns-restart.service cleaner IMHO.

File: nixos/modules/services/networking/cjdns.nix (hunk 3 above)
Comment (spacekitteh) [inline nixos/modules/services/networking/cjdns.nix]: Becuase cjdns requires the network to actually be online.

File: nixos/modules/services/networking/cjdns.nix (hunk 1 above)
Comment (spacekitteh) [inline nixos/modules/services/networking/cjdns.nix]: Whoops, I forgot to remove that line

File: nixos/modules/services/networking/cjdns.nix (hunk 3 above)
Comment (Mic92) [inline nixos/modules/services/networking/cjdns.nix]: From understanding this is the job of `Restart = "always";`, which does the same, right?

File: nixos/modules/services/networking/cjdns.nix [hunk 4]
```diff
@@ -204,11 +204,28 @@ in
...
     # networking.firewall.allowedUDPPorts = ...
-
+    systemd.services.cjdns-restart = {
+      description = "cjdns: restart after resuming from sleep, hibernation, hybrid-sleep and hybrid-suspend; and other failures";
+      wantedBy = ["sleep.target"];
+      bindsTo = ["network-online.target"];
+      after = ["sleep.target" "network-online.target"];
+
+      script = 
+        "
```
Comment (grahamc) [inline nixos/modules/services/networking/cjdns.nix]: Looks like this should either be a single-line string, or use
```
script = ''
  double single quotes
'';
```

**Review by grahamc — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/cjdns.nix (hunk 4 above)
Comment (spacekitteh) [inline nixos/modules/services/networking/cjdns.nix]: After rethinking I just deleted the -restart service altogether.

**Comment by grahamc** [comment]
Merged in 0ea07968588f172538647c2f28bef03fdc90f0b2, thank you for your effort and time on this :)

**Comment by spacekitteh** [comment]
\o/

**Pull request closed without merging**